### PR TITLE
Edit vNIC MAC address

### PR DESCRIPTION
**Pull request by skobyda**
Fixes https://github.com/cockpit-project/cockpit-machines/issues/40

Screenshot from 2021-04-20 13-40-19

### CODE REVIEW

File: cockpit-machines.spec.in [hunk 1]
```diff
@@ -57,7 +57,7 @@ appstream-util validate-relax --nonet %{buildroot}/%{_datadir}/metainfo/*
 
 %files
 %doc README.md
-%license LICENSE dist/index.js.LICENSE.txt
+%license LICENSE dist/LICENSE
```
Comment (KKoukiou) [inline cockpit-machines.spec.in:60]: Unrelated change.

File: src/components/vm/nics/nicEdit.jsx [hunk 2]
```diff
@@ -19,25 +19,23 @@
...
 const _ = cockpit.gettext;
 
-const NetworkMacRow = ({ network }) => {
+const NetworkMacRow = ({ mac, onValueChanged, idPrefix }) => {
     return (
-        <FormGroup fieldId='mac' label={_("MAC address")} hasNoPaddingTop>
-            <samp id='mac'>
-                {network.mac}
-            </samp>
+        <FormGroup fieldId="mac" label={_("MAC address")} hasNoPaddingTop>
```
Comment (KKoukiou) [inline src/components/vm/nics/nicEdit.jsx:35]: The fieldId needs to be the same with the id in the nested component for accessibility purposed.

File: src/libvirt-dbus.js [hunk 3]
```diff
@@ -1450,6 +1334,56 @@ export function changeNetworkAutostart(network, autostart, dispatch) {
...
+     * 2 -> VIR_DOMAIN_AFFECT_CONFIG
+     */
+    let flags = Enum.VIR_DOMAIN_AFFECT_CURRENT;
+    flags |= Enum.VIR_DOMAIN_AFFECT_CONFIG;
+
+    if (newMacAddress && newMacAddress !== macAddress) {
+        console.log("Thus Happend");
+        return detachIface(macAddress, connectionName, objPath, hotplug, persistent, dispatch)
+                .then(() => {
+                    console.log("This Happend");
```
Comment (KKoukiou) [inline src/libvirt-dbus.js:1362]: You forgot this :)

File: src/libvirt-dbus.js [hunk 4]
```diff
@@ -1450,6 +1334,56 @@ export function changeNetworkAutostart(network, autostart, dispatch) {
...
+    /*
+     * 0 -> VIR_DOMAIN_AFFECT_CURRENT
+     * 1 -> VIR_DOMAIN_AFFECT_LIVE
+     * 2 -> VIR_DOMAIN_AFFECT_CONFIG
+     */
+    let flags = Enum.VIR_DOMAIN_AFFECT_CURRENT;
+    flags |= Enum.VIR_DOMAIN_AFFECT_CONFIG;
+
+    if (newMacAddress && newMacAddress !== macAddress) {
+        console.log("Thus Happend");
```
Comment (KKoukiou) [inline src/libvirt-dbus.js:1359]: You forgot this.

File: src/libvirt-dbus.js [hunk 5]
```diff
@@ -1450,6 +1334,56 @@ export function changeNetworkAutostart(network, autostart, dispatch) {
...
+        return detachIface(macAddress, connectionName, objPath, hotplug, persistent, dispatch)
+                .then(() => {
+                    console.log("This Happend");
+                    console.log({ connectionName, vmId: objPath, mac: newMacAddress, permanent: true, hotplug: false, sourceType: networkType, source: networkSource, model: networkModel });
+                    return attachIface({ connectionName, vmId: objPath, mac: newMacAddress, permanent: persistent, hotplug, sourceType: networkType, source: networkSource, model: networkModel });
```
Comment (KKoukiou) [inline src/libvirt-dbus.js:1364]: Could you make the line collapse, it's too long, in my desktop screen it get's hidden :/

**Review by KKoukiou — CHANGES_REQUESTED** [review]

**Review by KKoukiou — APPROVED** [review]
Thanks for this feature. I just noriced one more problem, but I am happy if we solve it in a followup, I will leave it up to you:
* When the user edits the mac when the VM is running there is a UI glitch causing both interfaces to appear in the UI for a second or two
* Then if the user will try to edit again the MAC there will be an error.
Screen Shot 2021-04-22 at 18 22 06

**Comment by KKoukiou** [comment]
One failure in the last run that looks suspicious, let's retry: https://logs.cockpit-project.org/logs/pull-116-20210422-071806-fe74b4fa-rhel-9-0/log.html#48

**Comment by KKoukiou** [comment]
@skobyda can you please add a release note text for this?

**Comment by skobyda** [comment]
> @skobyda can you please add a release note text for this?

Done https://github.com/cockpit-project/cockpit-machines/pull/116#issue-619249868

**Comment by martinpitt** [comment]
@KKoukiou Your suspicion was right. Unfortunately we don't re-run these tests 3 times yet. I filed #127 about it, as it now blocks all c-machines PRs.

**Comment by KKoukiou** [comment]
> @KKoukiou Your suspicion was right. Unfortunately we don't re-run these tests 3 times yet. I filed #127 about it, as it now blocks all c-machines PRs.

Sad :( Will take a look and also for running these tests 3 times in your pR.

**Comment by lzap** [comment]
Thank you so much! This looks great.

**Pull request merged**